miden-lib/asm/faucets/basic.masm

```diff
@@ -0,0 +1,68 @@
+use.miden::sat::account
+use.miden::sat::faucet
+use.miden::sat::tx
+
+#! Distributes freshly minted fungible assets to the provided address.
+#! 
+#! Inputs: [ASSET, tag, RECIPIENT, ...]
+#! Outputs: [note_ptr, 0, 0, 0, 0, 0, 0, 0, 0, ...]
+#!
+#! - ASSET is the fungible asset to be received.
+#! - tag is the tag to be included in the note.
+#! - RECIPIENT is the recipient of the note, i.e., 
+#!   hash(hash(hash(serial_num, [0; 4]), script_hash), input_hash).
+#! - note_ptr is the pointer to the memory address in the kernel.
+#!   This cannot directly be accessed from another context.
+#!
+#! FAILS if: 
+#! - The asset being minted is not associated with the faucet the transaction is being executed
+#!   against.
+#! - The asset is not well formed.
+#! - The total issuance after minting is greater than the maximum amount 
+#!   allowed.
+export.distribute
+    # get max supply of this faucet. We assume max_supply is stored at pos 3 in memory slot 0
+    push.0 exec.account::get_item drop drop drop 
+    # => [max_supply, ASSET, tag, RECIPIENT, ...]
+    
+    # get total issuance of this faucet so far and add amount to be minted
+    dup.4 exec.faucet::get_total_issuance add
+    # => [(total_issuance + amount_to_be_minted), max_supply, ASSET, tag, RECIPIENT, ...]
```

Comment:
If `amount_to_be_minted` is very large `total_issuance + amount_to_be_minted` can overflow and become a small number. We should put extra check in here to prevent this from happening.

Reply:
I check for overflow now with 

```
# fails if total_issuance + amount < total_issuance (check overflow)
    dup movdn.3 lte assert
    # => [(total_issuance + amount), amount, max_supply, amount, tag, RECIPIENT, ...] 

    # fails if total_issuance + amount < amount (check overflow)
    dup movdn.2 lte assert
    # => [(total_issuance + amount), max_supply, amount, tag, RECIPIENT, ...] 
```

Is there a more efficient way to check this?

Reply:
I think this works (or at least I can't think of a better way right now).